modules/OFConnectionManager2/module/src/cxn_instance.c

```diff
@@ -0,0 +1,2122 @@
+                         cxn_state_names[old_state]);
+        }
+        break;
+    default:
+        break;
+    }
+
+    /* Exit conditions for old state */
+    switch (old_state) {
+    case CXN_S_INIT:  /* fall-through */
+    case CXN_S_CLOSING:
+        ind_soc_timer_event_unregister(cxn_state_timeout, (void *)cxn);
+        break;
+    case CXN_S_HANDSHAKING:
+        /* local connections do not have a timeout */
+        if (!CXN_LOCAL(cxn)) {
+            ind_soc_timer_event_unregister(cxn_state_timeout, (void *)cxn);
+        }
+        break;
+    case CXN_S_HANDSHAKE_COMPLETE:
+        if (!CXN_LOCAL(cxn) && cxn->aux_id == 0) {
+            AIM_SYSLOG_INFO("Disconnected from controller <ip-address>:<port>",
+                            "The switch disconnected from the specified controller.",
+                            "Disconnected from controller %s",
+                            cxn->desc);
+        }
+        break;
+    default:
+        break;
+    }
+
+    /* Change state of connection */
+    cxn->state = new_state;
+
+    /* External notification of state change before processing */
+    if ((old_state == CXN_S_HANDSHAKING &&
+         new_state == CXN_S_HANDSHAKE_COMPLETE) ||
+        (old_state == CXN_S_HANDSHAKE_COMPLETE &&
+         new_state == CXN_S_CLOSING)) {
+        ind_cxn_status_change(cxn);
+    }
+
+    /* Post-processing of state transition. */
+    switch (new_state) {
+    case CXN_S_INIT:
+        /* this should never happen */
+        AIM_DIE("Error: transitioned to INIT");
+        break;
+
+    case CXN_S_HANDSHAKING:
+        cxn->controller->fail_count = 0;
+        cxn_register_debug_counters(cxn);
+        ind_cxn_bundle_init(cxn);
+        cxn_send_hello(cxn);
+        if (CXN_LOCAL(cxn)) {
+            /* Recursive call; transition to connected */
+            cxn_state_set(cxn, CXN_S_HANDSHAKE_COMPLETE);
+        } else {
+            ind_soc_timer_event_register_with_priority(
+                cxn_state_timeout, (void *)cxn,
+                cxn_state_timeouts[new_state], IND_CXN_EVENT_PRIORITY);
+        }
+        break;
+
+    case CXN_S_HANDSHAKE_COMPLETE:
+        if (cxn->keepalive.period_ms > 0) {
+            /* Set up periodic echo request */
+            ind_soc_timer_event_register_with_priority(
+                periodic_keepalive, (void *)cxn,
+                cxn->keepalive.period_ms, IND_CXN_EVENT_PRIORITY);
+        }
+
+        if (!CXN_LOCAL(cxn) && cxn->aux_id == 0) {
+            AIM_SYSLOG_INFO("Connected to controller <ip-address>:<port>",
+                            "The switch successfully connected to the specified controller.",
+                            "Connected to controller %s",
+                            cxn->desc);
+        }
+        break;
+
+    case CXN_S_CLOSING:
+        if (cxn->keepalive.period_ms > 0) {
+            ind_soc_timer_event_unregister(periodic_keepalive, (void *)cxn);
+        }
+        ind_soc_timer_event_register_with_priority(
```

Comment:
If we don't time out from the closing state, then we have to handle all the bugs where the switch stops connecting to controller and just hangs.  This is OK if we are passing a build to QA and want to catch all these cases, but less optimal if we release with that code.  How often are we encountering leaks or strange behavior from timing out in the closing state?

BTW, calling cxn_state_set with the same state is a no-op.

As far as reusing the connection goes, we are protected against this case by the generation id.